senderworker/src/main/scala/com/gu/notifications/ec2worker/SenderWorker.scala

```diff
@@ -4,57 +4,91 @@ import com.amazonaws.services.lambda.runtime.events.SQSEvent.SQSMessage
     .build()
 
-  def pollQueue(queue: String, appName: String, handleChunkTokens: (SQSEvent, Context) => Unit): Future[Unit] = Future {
-    logger.info(s"About to poll queue - $appName")
-    val receiveMessages = sqsClient.receiveMessage(new ReceiveMessageRequest().withQueueUrl(queue).withWaitTimeSeconds(20)).getMessages.asScala.toList
-
-    def parsedReceivedMessages(message: List[Message]): SQSEvent = {
-      val parsedMessages = message.map(m => {
-        val sqsMessage = new SQSMessage()
-        sqsMessage.setBody(m.getBody)
-        sqsMessage.setAttributes(m.getAttributes)
-        sqsMessage.setReceiptHandle(m.getReceiptHandle)
-        sqsMessage
-      })
-
-      val event = new SQSEvent()
-      event.setRecords(parsedMessages.asJava)
-      event
+  class OpenConnection(queueName: String) {
+    val connectionFactory = new SQSConnectionFactory(
+      new ProviderConfiguration(),
+      sqsClient
+    )
+
+    val connection: SQSConnection = connectionFactory.createConnection()
+
+    val connectionSession: Session = connection.createSession(false, Session.CLIENT_ACKNOWLEDGE)
+
+    val consumer: MessageConsumer = connectionSession.createConsumer(connectionSession.createQueue(queueName))
+
+    connection.start()
+  }
+
+  class ReceiverCallback(queueUrl: String, handleChunkTokens: (SQSEvent, Context) => Unit) extends MessageListener {
+
+    def parsedReceivedMessages(message: jms.Message): Option[SQSEvent] = {
+
+     message match {
+        case textMessage: TextMessage => {
+          val sqsMessage = new SQSMessage()
+          sqsMessage.setBody(textMessage.getText)
+          sqsMessage.setAttributes(Map("SentTimestamp" -> textMessage.getJMSTimestamp.toString).asJava)
+          val event = new SQSEvent()
+          event.setRecords(List(sqsMessage).asJava)
+          Some(event)
+          }
+        case _ => None
+      }
     }
 
-    val parsedEvent: SQSEvent = parsedReceivedMessages(receiveMessages)
+    override def onMessage(message: jms.Message): Unit = {
+      println("JMS message received")
+      val parsedEvent = parsedReceivedMessages(message)
+      parsedEvent match {
+        case Some(event) => {
+          handleChunkTokens(event, null)
+          message.acknowledge()
+        }
+        case None => {
+          logger.error("Cannot parse message")
+        }
+      }
+    }
+  }
 
-    handleChunkTokens(parsedEvent, null)
+  def listenForMessages(queueName: String, queueUrl: String, appName: String, handleChunkTokens: (SQSEvent, Context) => Unit): Unit = {
+    logger.info(s"About to poll queue - $appName")
 
-    parsedEvent.getRecords.forEach(m => {
-      sqsClient.deleteMessage(queue, m.getReceiptHandle)
-      logger.info("Successfully deleted message")
-    })
+    val connectionCallback = new ReceiverCallback(queueUrl, handleChunkTokens)
 
-    logger.info(s"Finished polling $appName")
+    new OpenConnection(queueName) {
+      consumer.setMessageListener(connectionCallback)
+    }
```

Comment:
If a message is received from SQS, will the callback function be executed in the same thread or in another thread?  I just wonder that if there are multiple messages, will they be handled concurrently?

If the callback are executed in separate threads, is it possible for us to tune the size of thread pool?

Reply:
I have now made a change so the `onMessage` function returns a future, which means the messages are received asynchronously and using different threads. It is using the thread pool size of the lambda, which is currently 50. Do you think this should be more?

| Application Logs Before | Application Logs After |
| --- | --- |
|<img src="https://user-images.githubusercontent.com/19835654/202517083-c0fcfa4b-07b1-4d26-91e2-0921dc179be5.png" width="600px" />|<img src="https://user-images.githubusercontent.com/19835654/202519255-c7a8a2a7-203d-4d01-a87f-f05102f15650.png" width="600px" /> |

Reply:
It's a great question - in our current setup we define the ec2 instance type as [t4g small](https://aws.amazon.com/ec2/instance-types/?trk=adf88fbf-e631-4352-979f-e3812ceb7322&sc_channel=ps&s_kwcid=AL!4422!3!536451532736!e!!g!!aws%20ec2%20instance%20types&ef_id=Cj0KCQiAg_KbBhDLARIsANx7wAwMGv0grfOqMTLnqxh4RpgIcntayYhvgsJqWnTlxUA4-vMxsFbOZwUaAhnyEALw_wcB:G:s&s_kwcid=AL!4422!3!536451532736!e!!g!!aws%20ec2%20instance%20types), which has a standard vcpu of 2, but with burstable cpu up to 24.

I thought the general rule scala had was to create a thread pool per cpu, but in the lambda we bumped this above what vcpu we thought we'd have because of the relatively small amount of work we want each thread to do (mostly just making a http request, then receiving/processing the response).

The benefit of the ec2 model is that we can vertically scale, so I think it might be worth:
- keeping the default thread pool size as it is for now (if we start to see out of memory errors during testing we can always reduce the thread pool size)
-  if we see slow processing of messages, increase the default number of instances in our target group

Reply:
Thanks Frankie!  I guess it may be a parameter which needs tuning when we test it on CODE

Reply:
I agree.  I think there are multiple dimensions for tuning: 
- instance size
- no. of instances in ASG
- no. of threads etc